src/query/expression/src/values.rs

```diff
@@ -1830,22 +1778,76 @@ impl ColumnBuilder {
                 assert_eq!(fields.len(), value.len());
-                for (field, scalar) in fields.iter_mut().zip(value.iter()) {
-                    field.push(scalar.clone());
+                for (field, scalar) in fields.iter_mut().zip(value.into_iter()) {
+                    field.push(scalar);
                 }
             }
             (ColumnBuilder::Variant(builder), ScalarRef::Variant(value)) => {
-                builder.put_slice(value);
-                builder.commit_row();
+                VariantType::push_item(builder, value);
             }
             (ColumnBuilder::Geometry(builder), ScalarRef::Geometry(value)) => {
-                builder.put_slice(value);
-                builder.commit_row();
+                GeometryType::push_item(builder, value);
             }
             (builder, scalar) => unreachable!("unable to push {scalar:?} to {builder:?}"),
         }
     }
 
+    pub fn push_repeat(&mut self, item: &ScalarRef, n: usize) {
+        match (self, item) {
+            (ColumnBuilder::Null { len }, ScalarRef::Null)
+            | (ColumnBuilder::EmptyArray { len }, ScalarRef::EmptyArray)
+            | (ColumnBuilder::EmptyMap { len }, ScalarRef::EmptyMap) => *len += n,
+
+            (ColumnBuilder::Number(builder), ScalarRef::Number(value)) => {
+                builder.push_repeat(*value, n)
+            }
+            (ColumnBuilder::Decimal(builder), ScalarRef::Decimal(value)) => {
+                builder.push_repeat(*value, n)
+            }
+            (ColumnBuilder::Boolean(builder), ScalarRef::Boolean(value)) => {
+                BooleanType::push_item_repeat(builder, *value, n);
+            }
+            (ColumnBuilder::Timestamp(builder), ScalarRef::Timestamp(value)) => {
+                TimestampType::push_item_repeat(builder, *value, n);
+            }
+            (ColumnBuilder::Date(builder), ScalarRef::Date(value)) => {
+                DateType::push_item_repeat(builder, *value, n);
+            }
+            (ColumnBuilder::Binary(builder), ScalarRef::Binary(value)) => {
+                BinaryType::push_item_repeat(builder, *value, n);
+            }
+            (ColumnBuilder::String(builder), ScalarRef::String(value)) => {
+                StringType::push_item_repeat(builder, *value, n);
+            }
+            (ColumnBuilder::Array(builder), ScalarRef::Array(value))
+            | (ColumnBuilder::Map(builder), ScalarRef::Map(value)) => {
+                builder.push_repeat(value, n);
+            }
+            (ColumnBuilder::Bitmap(builder), ScalarRef::Bitmap(value)) => {
+                BitmapType::push_item_repeat(builder, value, n);
+            }
+            (ColumnBuilder::Nullable(builder), ScalarRef::Null) => {
+                NullableType::push_item_repeat(builder, None, n);
+            }
+            (ColumnBuilder::Nullable(builder), value) => {
+                builder.push_repeat(value.clone(), n);
```

Comment:
`ScalarRef` or `&ScalarRef`? I've spent quite a bit of time here, and still haven't found a complete solution to the problem

Reply:
ScalarRef is ok。